libV2/config/index.js

```diff
@@ -0,0 +1,223 @@
+const fs = require('fs');
+const path = require('path');
+const Joi = require('@hapi/joi');
+const { truthy } = require('../constants');
+
+const { envNamespace } = require('../constants');
+
+const configSchema = require('./schema');
+
+const _typeCasts = {
+    bool: val => truthy.includes(val.toLowerCase()),
+    int: val => parseInt(val, 10),
+    list: val => val.split(',').map(v => v.trim()),
+};
+
+function _definedInEnv(key) {
+    return process.env[`${envNamespace}_${key}`] !== undefined;
+}
+
+function _loadFromEnv(key, defaultValue, type) {
+    if (_definedInEnv(key)) {
+        const envKey = `${envNamespace}_${key}`;
+        const value = process.env[envKey];
+        if (type !== undefined) {
+            return _typeCasts[type](value);
+        }
+        return value;
+    }
+    return defaultValue;
+}
+
+const defaultConfigPath = path.join(__dirname, '../../config.json');
+
+class Config {
+    /**
+     * Returns a new Config instance merging the loaded config with the provided values.
+     * Passed values override loaded ones recursively.
+     *
+     * @param {object} overrides - an object using the same structure as the config file
+     * @returns {Config} - New Config instance
+     */
+    constructor(overrides) {
+        this._basePath = path.join(__dirname, '../../');
+        this._configPath = _loadFromEnv('CONFIG_FILE', defaultConfigPath);
+        this._defaultsPath = path.join(__dirname, 'defaults.json');
+
+        this.host = undefined;
+        this.port = undefined;
+
+        this.healthChecks = undefined;
+        this.logging = { level: 'debug', dumpLevel: 'error' };
+
+        this.redis = undefined;
+        this.warp10 = undefined;
+
+        // read config automatically
+        const loadedConfig = this._loadConfig();
+        let parsedConfig = Config._parseConfig(loadedConfig);
+        if (overrides && typeof overrides === 'object') {
+            parsedConfig = this._recursiveUpdate(parsedConfig, overrides);
+        }
+        Object.assign(this, parsedConfig);
+    }
+
+    static _readFile(path) {
+        try {
+            const data = fs.readFileSync(path, {
+                encoding: 'utf-8',
+            });
+            return JSON.parse(data);
+        } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error(`error reading config file at ${path}`, { error });
+            throw new Error(`could not parse config file: ${error.message}`);
```

Comment:
I added it here to make sure the something is written to the console in case of error as the config module is wired in pretty deeply and I don't want to accidentally silently eat the error

Reply:
Hmm, I guess that is fine but it will be pretty hard to silently eat all the errors that will come from this I think.
You could also just rethrow the error and keep the error log.